**Assignment 2**               ![Spark logo](https://spark.apache.org/images/spark-logo-trademark.png)
                                 
   Association Rule Mining Assignment.   
   I will answer a few questions using FP growth on a Retail data set. 
   
   
   We will start by setting and grouping the dataset by  'InvoiceNo' and will concatenate 'StockCode' data and create a .txt file that we use as training data.


In [1]:
import pandas as pd
df = pd.read_csv('Retail.csv', encoding = "ISO-8859-1")
s2 =df.groupby('InvoiceNo')['StockCode'].apply(','.join)
s2 = s2.reset_index()
s2[:5].to_csv('Retail.txt', columns = ["StockCode"], header=False, index=False)

In [2]:
raw_rdd = s2

I will run a count on the dataset

In [3]:
raw_rdd.count()

InvoiceNo    25900
StockCode    25900
dtype: int64

We will take the first 5 .

In [4]:
s2.head(5)

InvoiceNo                                          StockCode
0    536365      85123A,71053,84406B,84029G,84029E,22752,21730
1    536366                                        22633,22632
2    536367  84879,22745,22748,22749,22310,84969,22623,2262...
3    536368                            22960,22913,22912,22914
4    536369                                              21756

1) How many frequent itemsets were found with minimum support of .15?

In [7]:
from pyspark.mllib.fpm import FPGrowth
from pyspark import SparkContext
sc=SparkContext()

In [9]:
data=sc.textFile("Retail.txt")
transactions=data.map(lambda line: line.strip().split(','))

In [11]:
model=FPGrowth.train(transactions, minSupport=0.15, numPartitions=10)

In [14]:
freqItems=model.freqItemsets().collect()

In [15]:
len(freqItems)

4241

2) How many association rules were identified with a minimum confidence of .4?

In [17]:
freqItemsetsDict={}

In [20]:
for i in range(0,len(freqItems)):
    freqItemsetsDict[str(freqItems[i].items)]=freqItems[i].freq
    

In [21]:
miniConfidence=.4
rules=sorted(model._java_model.generateAssociationRules(miniConfidence).collect(),key=lambda x:x.confidence(),reverse =True)

In [23]:
len (rules)

25035

3) What is the support and confidence of the following association rule: ['21754', '22748'] => ['22745']? Go beyond reporting the values, but include an interpretation of these evaluation metrics.